File: src/i2c_interface.erl
%%%-------------------------------------------------------------------
%%% @author  drimtajm@github
%%% @copyright (C) 2013, Angela Johansson
%%% @doc
%%% An Erlang NIF that interfaces the I2C bus on the Raspberry Pi
%%%
%%% This interface uses the I2C ioctl driver which provides access to the
%%% I2C bus with a file descriptor. A precondition to using this software
%%% is a working I2C bus and an installed i2c-tools package.
%%%
%%% i2c_interface is free software:  you can redistribute it and/or modify
%%% it under the terms of the GNU Lesser General Public License as
%%% published by the Free Software Foundation, either version 3 of the
%%% License, or (at your option) any later version.
%%%
%%% i2c_interface is distributed in the hope that it will be useful,
%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%%% GNU Lesser General Public License for more details.
%%%
%%% You should have received a copy of the GNU Lesser General Public License
%%% along with this software.  If not, see <http://www.gnu.org/licenses/>.
%%%
%%% @end
%%% Created : 30 Jul 2013 by drimtajm
%%%-------------------------------------------------------------------
-module(i2c_interface).

-export([open_i2c_bus/1, close_i2c_bus/1]).
-export([read_i2c_raw_byte/1, read_i2c_raw_word/1,
	 read_i2c_raw_signed_word/1, read_i2c_smbus_byte/2,
	 read_i2c_smbus_word/2, read_i2c_smbus_signed_word/2]).
-export([write_i2c_smbus_byte/3, write_i2c_smbus_word/3]).

-define(nif_stub,
        erlang:nif_error({nif_not_loaded, module, ?MODULE, line, ?LINE})).

-ifndef(test).
-on_load(on_load/0).
-else.
-export([on_load/0]).
-endif.

-type file_descriptor()  :: pos_integer().
-type i2c_address()      :: pos_integer().
-type register_address() :: byte().
-type error_code()       :: atom().
-type word()             :: 0..65535.

on_load() ->
    PrivDir = case code:priv_dir(?MODULE) of
                  {error, bad_name} ->
                      EbinDir = filename:dirname(code:which(?MODULE)),
                      AppPath = filename:dirname(EbinDir),
                      filename:join(AppPath, "priv");
                  Path ->
                      Path
              end,
    Filename = filename:join(PrivDir, ?MODULE),
    ok = erlang:load_nif(Filename, 0).


-spec(open_i2c_bus(i2c_address()) ->
	     {ok, file_descriptor()} | {error, error_code()}).
%%% @doc This opens the I2C bus, connects to the device at the specified
%%%      address and returns a file handle to the open bus.
open_i2c_bus(I2CAddress) ->
    open_i2c_bus_nif(I2CAddress).

-spec(close_i2c_bus(file_descriptor()) ->
	     ok | {error, error_code()}).
%%% @doc This closes the I2C bus for a given file handle.
close_i2c_bus(FileHandle) ->
    close_i2c_bus_nif(FileHandle).

-spec(read_i2c_raw_byte(file_descriptor()) ->
	     {ok, byte()} | {error, error_code()}).
%%% @doc This reads a raw byte value from the I2C device.
read_i2c_raw_byte(FileHandle) ->
    read_i2c_raw_byte_nif(FileHandle).

-spec(read_i2c_raw_word(file_descriptor()) ->
	     {ok, word()} | {error, error_code()}).
%%% @doc This reads a raw word value from the I2C device.
read_i2c_raw_word(FileHandle) ->
    read_i2c_raw_word_nif(FileHandle).

-spec(read_i2c_raw_signed_word(file_descriptor()) ->
	     {ok, word()} | {error, error_code()}).
%%% @doc This reads a signed word value from the I2C device.
read_i2c_raw_signed_word(FileHandle) ->
    read_i2c_raw_signed_word_nif(FileHandle).

-spec(read_i2c_smbus_byte(file_descriptor(), register_address()) ->
	     {ok, byte()} | {error, error_code()}).
%%% @doc This reads a register value from the I2C device.
read_i2c_smbus_byte(FileHandle, Register) ->
    read_i2c_smbus_byte_nif(FileHandle, Register).

-spec(read_i2c_smbus_word(file_descriptor(), register_address()) ->
	     {ok, word()} | {error, error_code()}).
%%% @doc This reads a register value from the I2C device.
read_i2c_smbus_word(FileHandle, Register) ->
    read_i2c_smbus_word_nif(FileHandle, Register).

-spec(read_i2c_smbus_signed_word(file_descriptor(), register_address()) ->
	     {ok, word()} | {error, error_code()}).
%%% @doc This reads a signed register value from the I2C device.
read_i2c_smbus_signed_word(FileHandle, Register) ->
    read_i2c_smbus_signed_word_nif(FileHandle, Register).

-spec(write_i2c_smbus_byte(file_descriptor(), register_address(),
		     byte()) ->
	     ok | {error, error_code()}).
%%% @doc This writes a byte value to a register on the I2C device.
write_i2c_smbus_byte(FileHandle, Register, Value) ->
    write_i2c_smbus_byte_nif(FileHandle, Register, Value).

-spec(write_i2c_smbus_word(file_descriptor(), register_address(),
		     word()) ->
	     ok | {error, error_code()}).
%%% @doc This writes a word value to a register on the I2C device.
write_i2c_smbus_word(FileHandle, Register, Value) ->
    write_i2c_smbus_word_nif(FileHandle, Register, Value).

%%%%%%%%%%%%%%%
%% Define stubs for NIF functions

open_i2c_bus_nif(_I2CAddress)                            -> ?nif_stub.
read_i2c_raw_byte_nif(_FileHandle)                       -> ?nif_stub.
read_i2c_raw_word_nif(_FileHandle)                       -> ?nif_stub.
read_i2c_raw_signed_word_nif(_FileHandle)                -> ?nif_stub.
read_i2c_smbus_byte_nif(_FileHandle, _Register)          -> ?nif_stub.
read_i2c_smbus_word_nif(_FileHandle, _Register)          -> ?nif_stub.
read_i2c_smbus_signed_word_nif(_FileHandle, _Register)   -> ?nif_stub.
write_i2c_smbus_byte_nif(_FileHandle, _Register, _Value) -> ?nif_stub.
write_i2c_smbus_word_nif(_FileHandle, _Register, _Value) -> ?nif_stub.
close_i2c_bus_nif(_FileHandle)                           -> ?nif_stub.

%%
%%%%%%%%%%%%%%%

File: c_src/i2c_interface.c
// An Erlang NIF which interfaces the i2c driver exposed by ioctl. 
// Copyright (C) 2013 Angela Johansson
//
// This file is part of free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This i2c interface is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this code.  If not, see <http://www.gnu.org/licenses/>.

#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <sys/ioctl.h>

#ifdef __ARM_EABI__
// ARM architecture, we're probably on a Raspberry Pi. Include "real" I2C headers
#include <linux/i2c-dev.h>
#else
// Not on ARM, probably compiling on a PC. Include I2C stub headers
#include "i2c-stub.h"
#endif

#include "erl_nif.h"

static ERL_NIF_TERM atom_ok;
static ERL_NIF_TERM atom_error;
static ERL_NIF_TERM atom_other;

static int load(ErlNifEnv* env, void** priv_data, ERL_NIF_TERM load_info) {
    atom_ok    = enif_make_atom(env, "ok");
    atom_error = enif_make_atom(env, "error");
    atom_other = enif_make_atom(env, "other");
    return 0;
}

static ERL_NIF_TERM errno2atom(ErlNifEnv *env, const int error_code) {
  switch (error_code) {
  case EACCES: return enif_make_atom(env, "eacces");
  case EINVAL: return enif_make_atom(env, "einval");
  case ENOENT: return enif_make_atom(env, "enoent");
  case EBADF:  return enif_make_atom(env, "ebadf");
    // TODO: implement other error codes
  default:     return enif_make_tuple(env, 2, atom_other,
				      enif_make_int(env, error_code));
  }
}

static ERL_NIF_TERM make_error(ErlNifEnv *env, const int error_code) {
  return enif_make_tuple(env, 2, atom_error, errno2atom(env, error_code));
}

static ERL_NIF_TERM open_i2c_bus_nif(ErlNifEnv *env, int argc,
				     const ERL_NIF_TERM argv[]) {
  int file, address, result;
  if (!enif_get_int(env, argv[0], &address)) {
    return enif_make_badarg(env);
  }
#ifdef __ARM_EABI__
  // My Raspberry Pi uses I2C bus 0.
  // If necessary, change the line to match the bus number of _your_ system:
  char *filename = "/dev/i2c-0";
#else
  char *filename = "i2c-bus-stub";
#endif
  if ((file = open(filename, O_RDWR)) < 0) {
    result = errno;
    return make_error(env, result);
  }
  // The bus is open, now communicate to the device pointed out by address
  if (ioctl(file, I2C_SLAVE, address) < 0) {
    return make_error(env, errno);
  }
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, file));
}

static unsigned int swap_lower_bytes(const unsigned int val) {
  const unsigned int tmp = val & 0xFFFF;
  return ((tmp << 8) | (tmp >> 8)) & 0xFFFF;
}

static ERL_NIF_TERM read_i2c_raw_byte_nif(ErlNifEnv *env, int argc,
					  const ERL_NIF_TERM argv[]) {
  int file;
  int16_t result;
  uint8_t value;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((result = read(file, &value, 1)) != 1) {
    return make_error(env, errno);
  }
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, value));
}

static ERL_NIF_TERM read_i2c_raw_word_nif(ErlNifEnv *env, int argc,
					  const ERL_NIF_TERM argv[]) {
  int file;
  int32_t result;
  uint16_t value0;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((result = read(file, &value0, 2)) != 2) {
    return make_error(env, errno);
  }
  unsigned int value = swap_lower_bytes(value0);
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, value));
}

static ERL_NIF_TERM read_i2c_raw_signed_word_nif(ErlNifEnv *env, int argc,
						 const ERL_NIF_TERM argv[]) {
  int file;
  int32_t result;
  uint16_t value0;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((result = read(file, &value0, 2)) != 2) {
    return make_error(env, errno);
  }
  int value = (int16_t)swap_lower_bytes(value0);
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, value));
}

static ERL_NIF_TERM read_i2c_smbus_byte_nif(ErlNifEnv *env, int argc,
					    const ERL_NIF_TERM argv[]) {
  int file;
  unsigned int reg;
  int16_t result;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[1], &reg)) || (reg >= 256)) {
    return enif_make_badarg(env);
  }
  if ((result = i2c_smbus_read_byte_data(file, reg)) < 0) {
    return make_error(env, errno);
  }
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, result));
}

static ERL_NIF_TERM read_i2c_smbus_word_nif(ErlNifEnv *env, int argc,
					    const ERL_NIF_TERM argv[]) {
  int file;
  unsigned int reg;
  int32_t result;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[1], &reg)) || (reg >= 256)) {
    return enif_make_badarg(env);
  }
  if ((result = i2c_smbus_read_word_data(file, reg)) < 0) {
    return make_error(env, errno);
  }
  unsigned int value = swap_lower_bytes(result);
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, value));
}

static ERL_NIF_TERM read_i2c_smbus_signed_word_nif(ErlNifEnv *env, int argc,
						   const ERL_NIF_TERM argv[]) {
  int file;
  unsigned int reg;
  int32_t result;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[1], &reg)) || (reg >= 256)) {
    return enif_make_badarg(env);
  }
  if ((result = i2c_smbus_read_word_data(file, reg)) < 0) {
    return make_error(env, errno);
  }
  int value = (int16_t)swap_lower_bytes(result);
  return enif_make_tuple(env, 2, atom_ok, enif_make_int(env, value));
}

static ERL_NIF_TERM write_i2c_smbus_byte_nif(ErlNifEnv *env, int argc,
				       const ERL_NIF_TERM argv[]) {
  int file;
  unsigned int reg, value0;
  int32_t result;
  uint8_t value;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[1], &reg)) || (reg >= 256)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[2], &value0)) || (value0 >= 256)) {
    return enif_make_badarg(env);
  }
  value = value0;
  if ((result = i2c_smbus_write_byte_data(file, reg, value)) < 0) {
    return make_error(env, errno);
  }
  return atom_ok;
}

static ERL_NIF_TERM write_i2c_smbus_word_nif(ErlNifEnv *env, int argc,
				       const ERL_NIF_TERM argv[]) {
  int file;
  unsigned int reg, value0;
  int32_t result;
  uint16_t value;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[1], &reg)) || (reg >= 256)) {
    return enif_make_badarg(env);
  }
  if ((!enif_get_uint(env, argv[2], &value0)) || (value0 >= 65536)) {
    return enif_make_badarg(env);
  }
  value = swap_lower_bytes(value0);
  if ((result = i2c_smbus_write_word_data(file, reg, value)) < 0) {
    return make_error(env, errno);
  }
  return atom_ok;
}

static ERL_NIF_TERM close_i2c_bus_nif(ErlNifEnv *env, int argc,
				      const ERL_NIF_TERM argv[]) {
  int file, result;
  if (!enif_get_int(env, argv[0], &file)) {
    return enif_make_badarg(env);
  }
  if ((result = close(file)) < 0) {
    result = errno;
    return enif_make_tuple(env, 2, atom_error, errno2atom(env, result));
    //return enif_make_tuple(env, 2, atom_error, enif_make_int(env, result));  
  }
  return atom_ok;
}

static ErlNifFunc nif_funcs[] =
    {
      {"open_i2c_bus_nif",               1, open_i2c_bus_nif},
      {"read_i2c_raw_byte_nif",          1, read_i2c_raw_byte_nif},
      {"read_i2c_raw_word_nif",          1, read_i2c_raw_word_nif},
      {"read_i2c_raw_signed_word_nif",   1, read_i2c_raw_signed_word_nif},
      {"read_i2c_smbus_byte_nif",        2, read_i2c_smbus_byte_nif},
      {"read_i2c_smbus_word_nif",        2, read_i2c_smbus_word_nif},
      {"read_i2c_smbus_signed_word_nif", 2, read_i2c_smbus_signed_word_nif},
      {"write_i2c_smbus_byte_nif",       3, write_i2c_smbus_byte_nif},
      {"write_i2c_smbus_word_nif",       3, write_i2c_smbus_word_nif},
      {"close_i2c_bus_nif",              1, close_i2c_bus_nif}
    };

ERL_NIF_INIT(i2c_interface, nif_funcs, load, NULL, NULL, NULL)
